topdown/graphql.go

```diff
@@ -0,0 +1,192 @@
+		// This is wasteful, but a shorter path than requiring the caller
+		// to do it themselves immediately afterwards.
+		if err := result.UnmarshalJSON(val); err != nil {
+			return result, err
+		}
+	default:
+		return result, errGraphQLError(builtinName, "non-AST type fed into AST-to-JSON transformer.")
+	}
+	return result, nil
+}
+
+func builtinGraphQLParse(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Get the raw strings from each operand so that we can
+	// feed them to the GraphQL parser functions.
+	raw_query, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+	raw_schema, err := builtins.StringOperand(operands[1].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Generate ASTs/errors for the GraphQL schema and query.
+	schema, errQ := gqltop.LoadSchema(&gqlast.Source{Input: string(raw_schema)})
+	if errQ != nil {
+		return errQ.Unwrap()
+	}
+	query, errS := gqlparser.ParseQuery(&gqlast.Source{Input: string(raw_query)})
+	if errQ != nil {
+		return errS.Unwrap()
+	}
+
+	// Validate the query against the schema, erroring if there's an issue.
+	errs := gqlvalidator.Validate(schema, query)
+	if errs != nil {
+		return errGraphQLErrorList(ast.GraphQLParse.Name, errs)
+	}
+
+	// Transform the AST to JSON, then into an Object.
+	result, err := transformASTToTerm(ast.GraphQLParse.Name, query)
+	if err != nil {
+		return err
+	}
+
+	return iter(&result)
+}
+
+func builtinGraphQLParseQuery(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	raw, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Get the high-nested AST struct, along with any errors generated.
+	AST, errQ := gqlparser.ParseQuery(&gqlast.Source{Input: string(raw)})
+	if errQ != nil {
+		return errQ.Unwrap()
+	}
+
+	// Transform the AST to JSON, then into an Object.
+	result, err := transformASTToTerm(ast.GraphQLParse.Name, AST)
+	if err != nil {
+		return err
+	}
+
+	return iter(&result)
+}
+
+func builtinGraphQLParseSchema(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	raw, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Get the high-nested AST struct, along with any errors generated.
+	AST, errS := gqlparser.ParseSchema(&gqlast.Source{Input: string(raw)})
+	if errS != nil {
+		return errS.Unwrap()
+	}
+
+	// Transform the AST to JSON, then into an Object.
+	result, err := transformASTToTerm(ast.GraphQLParse.Name, AST)
+	if err != nil {
+		return err
+	}
+
+	return iter(&result)
+}
+
+func builtinGraphQLIsValid(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Get the raw strings from each operand so that we can
+	// feed them to the GraphQL parser functions.
+	raw_query, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+	raw_schema, err := builtins.StringOperand(operands[1].Value, 1)
```

Comment:
[style] camelCase is what you commonly do in Golang:
```suggestion
	rawSchema, err := builtins.StringOperand(operands[1].Value, 1)
```

Reply:
Thanks for the reminder! My history of working in C/Python where snake_case is the norm is coming back to bite me. :sweat_smile:

Reply:
This style issue should be resolved now. Thanks for catching it!